File: internal/service/metrics.go
package service

import (
	"fmt"
	"math"
	"strings"
	"time"

	cesModel "github.com/huaweicloud/huaweicloud-sdk-go-v3/services/ces/v1/model"
	gaussModel "github.com/huaweicloud/huaweicloud-sdk-go-v3/services/gaussdb/v3/model"

	cliModel "github.com/youweichen/taurusdb-cli/internal/model"
	"github.com/youweichen/taurusdb-cli/sdk"
)

// FetchInstanceMetrics queries Cloud Eye (CES) for a few key metrics that match
// the console "instance overview" style: CPU/Memory/SlowQueries/Connections.
//
// Notes:
// - GaussDB(for MySQL) uses namespace "SYS.GAUSSDB".
// - Dimensions are typically: gaussdb_mysql_instance_id, gaussdb_mysql_node_id.
// - Metric names are from the official GaussDB(for MySQL) monitoring doc.
func FetchInstanceMetrics(profile string, inst *gaussModel.MysqlInstanceInfoDetailUnifyStatus) (*cliModel.InstanceMetrics, error) {
	if inst == nil {
		return nil, fmt.Errorf("实例为空")
	}
	node := pickMasterNode(inst.Nodes)
	if node == nil || strings.TrimSpace(node.Id) == "" {
		return nil, fmt.Errorf("无法确定 master 节点 ID")
	}

	client, err := sdk.NewCESClient(profile)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	to := now.UnixMilli()
	// Fetch a small window and take the latest datapoint to avoid "aggregation not ready" empties.
	period := int32(300) // 5m
	window := 60 * time.Minute
	from := now.Add(-window).UnixMilli()

	dim0 := fmt.Sprintf("%s,%s", "gaussdb_mysql_instance_id", inst.Id)
	dim1 := fmt.Sprintf("%s,%s", "gaussdb_mysql_node_id", node.Id)

	ns := "SYS.GAUSSDB"
	out := &cliModel.InstanceMetrics{}
	var firstErr error

	cpu, err := fetchLatestPoint(client, sdk.MetricDataOptions{
		Namespace:     ns,
		MetricName:    "gaussdb_mysql001_cpu_util",
		Dim0:          dim0,
		Dim1:          &dim1,
		FromMs:        from,
		ToMs:          to,
		PeriodSeconds: period,
		Filter:        sdk.MetricFilterAverage,
	}, pickValue("average"))
	if err != nil && firstErr == nil {
		firstErr = err
	}
	out.CPUUtilPct = cpu

	mem, err := fetchLatestPoint(client, sdk.MetricDataOptions{
		Namespace:     ns,
		MetricName:    "gaussdb_mysql002_mem_util",
		Dim0:          dim0,
		Dim1:          &dim1,
		FromMs:        from,
		ToMs:          to,
		PeriodSeconds: period,
		Filter:        sdk.MetricFilterAverage,
	}, pickValue("average"))
	if err != nil && firstErr == nil {
		firstErr = err
	}
	out.MemUtilPct = mem

	slow, err := fetchLatestPoint(client, sdk.MetricDataOptions{
		Namespace:     ns,
		MetricName:    "gaussdb_mysql074_slow_queries",
		Dim0:          dim0,
		Dim1:          &dim1,
		FromMs:        from,
		ToMs:          to,
		PeriodSeconds: period,
		Filter:        sdk.MetricFilterSum,
	}, pickValue("sum"))
	if err != nil && firstErr == nil {
		firstErr = err
	}
	out.SlowQueries = slow

	conn, err := fetchLatestPoint(client, sdk.MetricDataOptions{
		Namespace:     ns,
		MetricName:    "gaussdb_mysql006_conn_count",
		Dim0:          dim0,
		Dim1:          &dim1,
		FromMs:        from,
		ToMs:          to,
		PeriodSeconds: period,
		Filter:        sdk.MetricFilterAverage,
	}, pickValue("average"))
	if err != nil && firstErr == nil {
		firstErr = err
	}
	out.ConnCount = conn

	return out, firstErr
}

func pickMasterNode(nodes *[]gaussModel.MysqlInstanceNodeInfo) *gaussModel.MysqlInstanceNodeInfo {
	if nodes == nil || len(*nodes) == 0 {
		return nil
	}
	for i := range *nodes {
		n := &(*nodes)[i]
		if n.Type != nil && strings.ToLower(strings.TrimSpace(*n.Type)) == "master" {
			return n
		}
	}
	return &(*nodes)[0]
}

type datapointValuePicker func(dp cesModel.Datapoint) *float64

func pickValue(filter string) datapointValuePicker {
	switch strings.ToLower(filter) {
	case "max":
		return func(dp cesModel.Datapoint) *float64 { return dp.Max }
	case "min":
		return func(dp cesModel.Datapoint) *float64 { return dp.Min }
	case "sum":
		return func(dp cesModel.Datapoint) *float64 { return dp.Sum }
	case "variance":
		return func(dp cesModel.Datapoint) *float64 { return dp.Variance }
	default:
		return func(dp cesModel.Datapoint) *float64 { return dp.Average }
	}
}

func fetchLatestPoint(client *sdk.CESClient, opts sdk.MetricDataOptions, pick datapointValuePicker) (*cliModel.MetricPoint, error) {
	resp, err := client.ShowMetricData(opts)
	if err != nil {
		return nil, err
	}
	if resp == nil || resp.Datapoints == nil || len(*resp.Datapoints) == 0 {
		return nil, nil
	}

	var best *cesModel.Datapoint
	for i := range *resp.Datapoints {
		dp := &(*resp.Datapoints)[i]
		if best == nil || dp.Timestamp > best.Timestamp {
			best = dp
		}
	}
	if best == nil {
		return nil, nil
	}
	vp := pick(*best)
	if vp == nil {
		return nil, nil
	}
	unit := ""
	if best.Unit != nil {
		unit = *best.Unit
	}
	v := *vp
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return nil, nil
	}
	return &cliModel.MetricPoint{
		Value:       v,
		Unit:        unit,
		TimestampMs: best.Timestamp,
	}, nil
}

File: src/api/metrics.ts
import { makeClient } from "../http/client.js";
import type {
  TaurusConfig,
  Instance,
  InstanceNode,
  InstanceMetrics,
  MetricPoint,
} from "../types/index.js";

async function showMetricData(
  cfg: TaurusConfig,
  opts: {
    namespace: string;
    metricName: string;
    dim0: string;
    dim1?: string;
    fromMs: number;
    toMs: number;
    periodSeconds: number;
    filter: "average" | "max" | "min" | "sum";
  },
): Promise<
  Array<{
    average?: number;
    max?: number;
    min?: number;
    sum?: number;
    timestamp: number;
    unit?: string;
  }>
> {
  const client = makeClient(cfg);
  const projectId = cfg.project_id!.trim();
  const query: Record<string, string> = {
    namespace: opts.namespace,
    metric_name: opts.metricName,
    "dim.0": opts.dim0,
    from: String(opts.fromMs),
    to: String(opts.toMs),
    period: String(opts.periodSeconds),
    filter: opts.filter,
  };
  if (opts.dim1?.trim()) query["dim.1"] = opts.dim1.trim();
  const resp = await client.request<{
    datapoints?: Array<{
      average?: number;
      max?: number;
      min?: number;
      sum?: number;
      timestamp: number;
      unit?: string;
    }>;
  }>("ces", "GET", `/V1.0/${projectId}/metric-data`, query);
  return resp.datapoints ?? [];
}

function pickMasterNode(inst: Instance): InstanceNode | null {
  const nodes = inst.nodes ?? [];
  if (nodes.length === 0) return null;
  const master = nodes.find(
    (n) => (n.type ?? "").trim().toLowerCase() === "master",
  );
  return master ?? nodes[0];
}

function pickDatapointValue(
  dp: {
    average?: number;
    max?: number;
    min?: number;
    sum?: number;
    timestamp: number;
    unit?: string;
  },
  filter: "average" | "max" | "min" | "sum",
): number | undefined {
  switch (filter) {
    case "max":
      return dp.max;
    case "min":
      return dp.min;
    case "sum":
      return dp.sum;
    default:
      return dp.average;
  }
}

async function fetchLatestMetricPoint(
  cfg: TaurusConfig,
  opts: {
    namespace: string;
    metricName: string;
    dim0: string;
    dim1?: string;
    fromMs: number;
    toMs: number;
    periodSeconds: number;
    filter: "average" | "max" | "min" | "sum";
  },
): Promise<MetricPoint | undefined> {
  const datapoints = await showMetricData(cfg, opts);
  if (datapoints.length === 0) return undefined;
  const latest = datapoints.reduce((best, cur) =>
    cur.timestamp > best.timestamp ? cur : best,
  );
  const value = pickDatapointValue(latest, opts.filter);
  if (value === undefined || Number.isNaN(value) || !Number.isFinite(value))
    return undefined;
  return {
    value,
    unit: (latest.unit ?? "").trim(),
    timestamp_ms: latest.timestamp,
  };
}

export async function fetchInstanceMetrics(
  cfg: TaurusConfig,
  inst: Instance,
): Promise<InstanceMetrics> {
  const master = pickMasterNode(inst);
  if (!master?.id?.trim()) {
    throw new Error("无法确定 master 节点 ID");
  }

  const now = Date.now();
  const periodSeconds = 300;
  const fromMs = now - 60 * 60 * 1000;
  const dim0 = `gaussdb_mysql_instance_id,${inst.id}`;
  const dim1 = `gaussdb_mysql_node_id,${master.id}`;
  const ns = "SYS.GAUSSDB";

  const metrics: InstanceMetrics = {};
  metrics.cpu_util_pct = await fetchLatestMetricPoint(cfg, {
    namespace: ns,
    metricName: "gaussdb_mysql001_cpu_util",
    dim0,
    dim1,
    fromMs,
    toMs: now,
    periodSeconds,
    filter: "average",
  });
  metrics.mem_util_pct = await fetchLatestMetricPoint(cfg, {
    namespace: ns,
    metricName: "gaussdb_mysql002_mem_util",
    dim0,
    dim1,
    fromMs,
    toMs: now,
    periodSeconds,
    filter: "average",
  });
  metrics.slow_queries = await fetchLatestMetricPoint(cfg, {
    namespace: ns,
    metricName: "gaussdb_mysql074_slow_queries",
    dim0,
    dim1,
    fromMs,
    toMs: now,
    periodSeconds,
    filter: "sum",
  });
  metrics.conn_count = await fetchLatestMetricPoint(cfg, {
    namespace: ns,
    metricName: "gaussdb_mysql006_conn_count",
    dim0,
    dim1,
    fromMs,
    toMs: now,
    periodSeconds,
    filter: "average",
  });
  return metrics;
}